OpenRA.Mods.Common/Warheads/DamageWarhead.cs

```diff
@@ -63,5 +74,27 @@ public virtual void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> damag
 			var damage = Util.ApplyPercentageModifiers(Damage, damageModifiers.Append(DamageVersus(victim)));
 			victim.InflictDamage(firedBy, new Damage(damage, DamageTypes));
 		}
+
+		public virtual void DoImpact(Actor victim, Actor firedBy, HitShapeInfo shape, IEnumerable<int> damageModifiers)
+		{
+			if (!IsValidAgainst(victim, firedBy))
+				return;
+
+			var damage = Util.ApplyPercentageModifiers(Damage, damageModifiers.Append(DamageVersus(victim, shape)));
+			victim.InflictDamage(firedBy, new Damage(damage, DamageTypes));
+		}
+
+		static bool ValidArmorType(string armor, HashSet<string> armorTypes)
+		{
+			if (armorTypes.Count == 0)
+				return true;
+
+			// PERF: Avoid LINQ.
+			foreach (var at in armorTypes)
+				if (armor.StartsWith(at))
```

Comment:
Why are you using `StartsWith`? This should also get a `StringComparison.Ordinal`.